odl/discr/tensor_ops.py

```diff
@@ -0,0 +1,610 @@
+
+        ``||F(x)|| = max_j( w_j * |F_j(x)| )``
+
+    for ``p = inf``, where ``F`` is a vector field. This implies that
+    the `Operator.domain` is a power space of a discretized function
+    space. For example, if ``X`` is a `DiscreteLp` space, then
+    ``ProductSpace(X, d)`` is a valid domain for any positive integer
+    ``d``.
+    """
+
+    def __init__(self, vfspace, exponent=None, weight=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        vfspace : `ProductSpace` of `Discretization`
+            Space of vector fields on which the operator acts.
+            It has to be a product space of identical discretized
+            function spaces, i.e. a power space. Its
+            `ProductSpace.shape` must have length 1.
+        exponent : non-zero `float`, optional
+            Exponent of the norm in each point. Values between
+            0 and 1 are currently not supported due to numerical
+            instability.
+            Default: ``domain.exponent``
+        weight : `array-like` or `float`, optional
+            Weighting array or constant for the norm. If an array is
+            given, its length must be equal to ``domain.size``, and
+            all entries must be positive. A provided constant must be
+            positive.
+            By default, the weights are is taken from
+            ``domain.weighting``. Note that this excludes unusual
+            weightings with custom inner product, norm or dist.
+
+        Examples
+        --------
+        We make a tiny vector field space in 2D and create the
+        standard point-wise norm operator on that space. The operator
+        maps a vector field to a scalar function:
+
+        >>> import odl
+        >>> spc = odl.uniform_discr([-1, -1], [1, 1], (2, 2))
+        >>> vfspace = odl.ProductSpace(spc, 2)
+        >>> pw_norm = PointwiseNorm(vfspace)
+        >>> pw_norm.range == spc
+        True
+
+        Now we can calculate the 2-norm in each point:
+
+        >>> x = vfspace.element([[[1, 2], [0, 3]],
+        ...                      [[0, 0], [2, 4]]])
+        >>> print(pw_norm(x))
+        [[1.0, 2.0],
+         [2.0, 5.0]]
+
+        We can change the exponent either in the vector field space
+        or in the operator directly:
+
+        >>> vfspace = odl.ProductSpace(spc, 2, exponent=1)
+        >>> pw_norm = PointwiseNorm(vfspace)
+        >>> print(pw_norm(x))
+        [[1.0, 2.0],
+         [2.0, 7.0]]
+        >>> vfspace = odl.ProductSpace(spc, 2)
+        >>> pw_norm = PointwiseNorm(vfspace, exponent=1)
+        >>> print(pw_norm(x))
+        [[1.0, 2.0],
+         [2.0, 7.0]]
+        """
+        if not isinstance(vfspace, ProductSpace):
+            raise TypeError('vector field space {!r} is not a ProductSpace '
+                            'instance.'.format(vfspace))
+        super().__init__(domain=vfspace, range=vfspace[0], linear=False)
+
+        if len(self.domain.shape) != 1:
```

Comment:
Currently nothing. Perhaps it should be a comment for later, when (perhaps) product spaces with more than 1-dimensional shapes are allowed.